lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,450 @@
+            |> Nx.put_slice([j, 0], Nx.reshape(infinities, {1, n}))
+            |> Nx.put_slice([0, j], Nx.reshape(infinities, {n, 1}))
+
+          # Update pointers
+          indices = [i, j] |> Nx.stack() |> Nx.new_axis(-1)
+          pointers = Nx.indexed_put(pointers, indices, Nx.stack([c, c]))
+
+          {clades, count + 1, pointers, pairwise, diss, sizes, links}
+        end
+      end
+
+    {clades, count, pointers, pairwise, diss, sizes}
+  end
+
+  defnp find_clade(pointers, i) do
+    {i, _, _} =
+      while {_current = i, next = pointers[i], pointers}, next != -1 do
+        {next, pointers[next], pointers}
+      end
+
+    i
+  end
+
+  # Dissimilarity functions
+
+  defnp pairwise_euclidean(%Nx.Tensor{} = x) do
+    x |> pairwise_euclidean_sq() |> Nx.sqrt()
+  end
+
+  defnp pairwise_euclidean_sq(%Nx.Tensor{} = x) do
+    sq = Nx.sum(x ** 2, axes: [1], keep_axes: true)
+    sq + Nx.transpose(sq) - 2 * Nx.dot(x, [1], x, [1])
+  end
+
+  # Dissimilarity update functions
+
+  defnp average(dac, dbc, _dab, sa, sb, _sc),
+    do: (sa * dac + sb * dbc) / (sa + sb)
+
+  # defnp centroid(dac, dbc, dab, sa, sb, _sc),
+  #   do: Nx.sqrt((sa * dac + sb * dbc) / (sa + sb) - sa * sb * dab / (sa + sb) ** 2)
+
+  defnp complete(dac, dbc, _dab, _sa, _sb, _sc),
+    do: Nx.max(dac, dbc)
+
+  # defnp median(dac, dbc, dab, _sa, _sb, _sc),
+  #   do: Nx.sqrt(dac / 2 + dbc / 2 - dab / 4)
+
+  defnp single(dac, dbc, _dab, _sa, _sb, _sc),
+    do: Nx.min(dac, dbc)
+
+  defnp ward(dac, dbc, dab, sa, sb, sc),
+    do: Nx.sqrt(((sa + sc) * dac + (sb + sc) * dbc - sc * dab) / (sa + sb + sc))
+
+  defnp weighted(dac, dbc, _dab, _sa, _sb, _sc),
+    do: (dac + dbc) / 2
+
+  # Cluster functions
+
+  @cluster_opts_schema [
+    by: [
+      type: :non_empty_keyword_list,
+      required: true,
+      keys: [
+        height: [
+          type: :float,
+          doc: "Height of the dendrogram to use as the split point for clusters."
+        ],
+        num_clusters: [
+          type: :pos_integer,
+          doc: "Number of clusters to form."
+        ]
+      ],
+      doc: """
+      How to select which clades from the dendrogram should form the final clusters.
+      Must provide either a height or a number of clusters.
+      """
+    ]
+  ]
+  @doc """
+  Cluster the dataset using the clades formed by `fit/2`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@cluster_opts_schema)}
+
+  ## Return values
+
+  Returns a `Nx.Tensor` with shape `{model.num_points}` and values `0..(k - 1)` where `k` is the
+  number of clusters formed.
+  The `i`th element of the result tensor is the label of the `i`th data point's cluster.
+  (Cluster labels are arbitrary, but deterministic.)
+
+  ## Examples
+
+      iex> data = Nx.tensor([[2], [7], [9], [0], [3]])
+      iex> model = Hierarchical.fit(data)
+      iex> Hierarchical.cluster(model, by: [num_clusters: 3])
+      Nx.tensor([0, 1, 1, 2, 0])
+  """
+  deftransform cluster(%__MODULE__{} = model, opts \\ []) do
```

Comment:
I went with `cluster/2` instead of `predict/2`. It serves a similar role as `predict/2` would, but it can only be used on the original data. There's no way to take the output of `fit/2` and apply it to new data.

(In that vein, `fit/2` may also be a misnomer.)

Reply:
@billylanchantin how is it called in scikit-learn? I would generally prefer to avoid introducing new names. If this is a single operations, perhaps we should use `fit_predict` instead? /cc @msluszniak

Reply:
Yes, `fit_predict` sounds good if it already returns the clustering

Reply:
Ah yeah they have `fit` and `fit_predict`. I'm good with that! @msluszniak You'll definitely want to take a pass and look for API consistency.

> Yes, `fit_predict` sounds good if it already returns the clustering

I actually had that first but switched. I'll switch back.

Reply:
[52b1e78](https://github.com/elixir-nx/scholar/pull/187/commits/52b1e78cf38cfd62d4e352c3ffb22cb9e528f556)

Reply:
Overtaken by: https://github.com/elixir-nx/scholar/pull/187#discussion_r1404707992